### PR TITLE
Feat/remove flexible booking option

### LINKED ISSUE #140: Remove flexible bookings
## Context

We want to be able to track stays better, therefore we will remove the option to make your booking an entirely _flexible_ booking, but will introduce a combination of having to give a date range as in `start_date` and `end_date` and an additional checkbox for `flexible`.

- Remove logic for a booking to be one or the other
- Make possible that a booking can be flexible, but it still needs a start date
- Display travel dates and if flexible has been selected too, also display flexible (bookings show)

Yes

### PR DESCRIPTION
**Pull request by lisbethpurrucker**
Issue number: resolves #140 

---

### Description

We removed the option for a booking to only be flexible, so from now on all bookings need a start and end date but can also be flexible.

For old bookings, we added a context when updating them (skipping validation) so that the crucial actions can still be performed successfully.

### Checklist

- [x] Code compiles correctly
- [x] Created tests which fail without the change (if possible)
- [x] All tests passing
- [x] Rubocop passes

### CODE REVIEW

File: app/controllers/concerns/bookings_concern.rb [hunk 1]
```diff
@@ -26,23 +26,29 @@ def post_update(booking)
...
 
   def decline_booking(booking, message = nil)
     raise Exceptions::ForbiddenError, 'Only the host can decline a booking' unless requested_by_host?(booking)
 
-    return unless booking.declined!
+    booking.status = :declined
+
+    return unless booking.save!(context: :change_status)
+
+    # flash[:error] = 'Booking could not be confirmed. Please ask your guest to specify dates or reach out for further assistance.'
```
Comment (papotte) [inline app/controllers/concerns/bookings_concern.rb:51]: What is this for?

File: test/controllers/concerns/bookings_concern_test.rb [hunk 2]
```diff
@@ -52,8 +52,11 @@ def setup
   end
 
   test 'should accept and send email' do
+    @booking.update(start_date: nil, end_date: nil)
     @booking.status = :pending
-    @booking.save!
+    @booking.save!(context: :change_status)
+
+    p @booking
```
Comment (papotte) [inline test/controllers/concerns/bookings_concern_test.rb:59]: Is this for logging? If you want to keep it, use Rails.logger.{info,debug,error}

File: test/controllers/concerns/bookings_concern_test.rb [hunk 3]
```diff
@@ -62,8 +65,9 @@ def setup
   end
 
   test 'should cancel request as guest' do
+    @booking.update(start_date: nil, end_date: nil)
```
Comment (papotte) [inline test/controllers/concerns/bookings_concern_test.rb:68]: Let's create two cases for each of these, one where the dates exist and another one where it doesn't.

**Review by papotte — CHANGES_REQUESTED** [review]
Please add more tests! At the very least I'm missing editing a request that is flexible (both the fail and the success case), editing a request that's complete, change of status for requests that are and are not flexible and creating a request without dates.

File: app/controllers/concerns/bookings_concern.rb (hunk 1 above)
Comment (lisbethpurrucker) [reply app/controllers/concerns/bookings_concern.rb:51]: oops, forgot to delete!

File: test/controllers/concerns/bookings_concern_test.rb (hunk 2 above)
Comment (lisbethpurrucker) [reply test/controllers/concerns/bookings_concern_test.rb:59]: and also forgot to delete...

File: app/assets/stylesheets/components/_bookings.scss [hunk 4]
```diff
@@ -97,6 +97,7 @@
 
 		i {
 			font-size: 16px;
+			width: 16px;
```
[the remark this answers is not in the document]
Comment (lisbethpurrucker) [reply app/assets/stylesheets/components/_bookings.scss:100]: I would like to show you in person, but there is a slight difference and they are not aligned perfectly without the `width` property. It's the font awesome icons (which when you hover are 14px width) vs. our own (16px width).

**Review by papotte — APPROVED** [review]

**Pull request merged**